internal/code-generator/generator/heuristics_test.go

```diff
@@ -0,0 +1,129 @@
+package generator
+
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"testing"
+	"text/template"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestYamlParsing(t *testing.T) {
+	heuristics, err := parseYaml("test_files/heuristics.yml")
+
+	require.NoError(t, err)
+	assert.NotNil(t, heuristics)
+
+	// extensions
+	require.NotNil(t, heuristics.Disambiguations)
+	assert.Equal(t, 4, len(heuristics.Disambiguations))
+	assert.Equal(t, 2, len(heuristics.Disambiguations[0].Extensions))
+
+	rules := heuristics.Disambiguations[0].Rules
+	assert.Equal(t, 2, len(rules))
+	require.Equal(t, "Objective-C", rules[0].Languages[0])
+	assert.Equal(t, 1, len(rules[0].Pattern))
+
+	rules = heuristics.Disambiguations[1].Rules
+	assert.Equal(t, 3, len(rules))
+	require.Equal(t, "Forth", rules[0].Languages[0])
+	require.Equal(t, 2, len(rules[0].Pattern))
+
+	rules = heuristics.Disambiguations[2].Rules
+	assert.Equal(t, 3, len(rules))
+	require.Equal(t, "Unix Assembly", rules[1].Languages[0])
+	require.NotNil(t, rules[1].And)
+	assert.Equal(t, 2, len(rules[1].And))
+	require.NotNil(t, rules[1].And[0].NegativePattern)
+	assert.Equal(t, "np", rules[1].And[0].NegativePattern)
+
+	rules = heuristics.Disambiguations[3].Rules
+	assert.Equal(t, 1, len(rules))
+	assert.Equal(t, "Linux Kernel Module", rules[0].Languages[0])
+	assert.Equal(t, "AMPL", rules[0].Languages[1])
+
+	// for i, rule := range rules {
+	// 	t.Logf("\n\t Forth rule%d: %q\n", i, rule)
+	// }
+
+	// named_patterns
+	require.NotNil(t, heuristics.NamedPatterns)
+	assert.Equal(t, 2, len(heuristics.NamedPatterns))
+	assert.Equal(t, 1, len(heuristics.NamedPatterns["fortran"]))
+	assert.Equal(t, 2, len(heuristics.NamedPatterns["cpp"]))
+}
+
+func TestSingleRuleLoading(t *testing.T) {
+	//given
```

Comment:
I'm not sure what the comments here mean. (But they are missing a leading space)